File: Assignments/6358289 Assignment 2/Deque.java
/**
 * The class Deque implements a double-ended queue with a doubly linked list.
 * The list uses a header and a trailer (dummy) nodes.
 *
 * @author Sebastian Nunez (6358289) - COP3530 U01 84883
 */
public class Deque {
    private int count; //number of elements in the deque
    private Node back; //points to the item in the back
    private Node front; //points to the item in the front

    /**
     * Default constructor. Sets this object as an empty deque.
     *
     */
    public Deque() {
        front = new Node();
        back = new Node();
        front.setNext(back);
        back.setPrev(front);
        count = 0;
    }

    /**
     * Adds new element to the back end of the deque. The method takes O(1)
     * time.
     *
     * @param x new element to be added to the deque.
     */
    public void addToBack(int x) {
        // new node
        Node ptr = new Node();
        ptr.setInfo(x);

        // adjust the new prev node to point to the new node (and vice versa )
        Node prev = back.getPrev();
        ptr.setPrev(prev);
        prev.setNext(ptr);

        // adjust the new node to point to the tail (and vise versa)
        ptr.setNext(back);
        back.setPrev(ptr);

        count++;
    }

    /**
     * Adds new element to the front end of the deque. The method takes O(1)
     * time.
     *
     * @param x new element to be added to the deque.
     */
    public void addToFront(int x) {
        // new node
        Node ptr = new Node();
        ptr.setInfo(x);

        // adjust the new next node to point to the new node (and vice versa )
        Node next = front.getNext();
        ptr.setNext(next);
        next.setPrev(ptr);

        // adjust the head to point to the new node (and vice versa)
        front.setNext(ptr);
        ptr.setPrev(front);

        count++;
    }

    /**
     * Retrieves element on the back end of the deque. The method takes O(1)
     * time.
     *
     * @return operation is successful: valid = true and item = element on the
     * back end; operation is unsuccessful (i.e. empty deque): valid = false and
     * item = dummy value
     */
    public DequeItem getBack() {
        if (isEmpty())
            return new DequeItem();

        Node last = back.getPrev();

        return new DequeItem(true, last.getInfo());
    }

    /**
     * Retrieves element on the front end of the deque. The method takes O(1)
     * time.
     *
     * @return operation is successful: valid = true and item = element on the
     * front end; operation is unsuccessful (i.e. empty deque): valid = false and
     * item = dummy value
     */
    public DequeItem getFront() {
        if (isEmpty())
            return new DequeItem();

        Node first = front.getNext();

        return new DequeItem(true, first.getInfo());
    }

    /**
     * Determines if deque is empty. The method takes O(1) time.
     *
     * @return true if deque contains no elements, false otherwise.
     */
    public boolean isEmpty() {
        return count == 0;
    }

    /**
     * Removes element on the back end of the deque. The method takes O(1) time.
     *
     * @return false if removal cannot be performed (i.e. the deque is empty),
     * true otherwise
     */
    public boolean removeBack() {
        if (isEmpty())
            return false;

        Node removed = back.getPrev(); // node to be removed
        Node prev = removed.getPrev(); // new previous node

        // new previous points to the back (and vice versa)
        prev.setNext(back);
        back.setPrev(prev);

        count--;
        return true;
    }

    /**
     * Removes element on the front end of the deque. The method takes O(1)
     * time.
     *
     * @return false if removal cannot be performed (i.e. the deque is empty),
     * true otherwise
     */
    public boolean removeFront() {
        if (isEmpty())
            return false;

        Node removed = front.getNext(); // node to be removed
        Node next = removed.getNext(); // new next node

        // the head now points to the new node (and vice versa)
        front.setNext(next);
        next.setPrev(front);

        count--;
        return true;
    }

    /**
     * Constructs a String description of the deque.
     *
     * @return String containing the deque elements.
     */
    public String toString() {
        String str = "";

        Node current = front.getNext();
        for (int i = 0; i < count - 1; i++) {
            str += current.getInfo() + ", ";
            current = current.getNext();
        }

        if (count != 0)
            return "Deque: [" + str + back.getPrev().getInfo() + "]";
        else
            return "Deque: []";
    }
}

File: Python/Deque.py
def main():
    deque = Deque()

    print(f'is_empty(): {deque.is_empty()}')
    print(deque)

    deque.append(4)
    deque.append(5)
    deque.appendleft(3)
    deque.appendleft(2)
    print(deque)

    print(f'Pop: {deque.pop()}')
    print(f'Pop: {deque.pop()}')
    print(f'Popleft: {deque.popleft()}')
    print(deque)

    print(f'Popleft: {deque.popleft()}')
    print(deque)
    print(f'Pop: {deque.pop()}')


class Node:
    def __init__(self, data: int = None):
        self.data = data
        self.next = None
        self.prev = None


class Deque:
    def __init__(self):
        self.head = Node()
        self.tail = Node()

        self.head.next = self.tail
        self.tail.prev = self.head

        self.size = 0

    def append(self, data: int) -> None:
        new_node = Node(data)
        prev = self.tail.prev

        new_node.prev = prev
        prev.next = new_node

        new_node.next = self.tail
        self.tail.prev = new_node

        self.size += 1

    def appendleft(self, data: int) -> None:
        new_node = Node(data)
        next = self.head.next

        new_node.next = next
        next.prev = new_node

        self.head.next = new_node
        new_node.prev = self.head

        self.size += 1

    def pop(self) -> int:
        if self.is_empty():
            print('deque is empty!')
            return None

        removed = self.tail.prev
        last = removed.prev

        last.next = self.tail
        self.tail.prev = last

        self.size -= 1
        return removed.data

    def popleft(self) -> int:
        if self.is_empty():
            print('deque is empty!')
            return None

        removed = self.head.next
        first = removed.next

        self.head.next = first
        first.prev = self.head

        self.size -= 1
        return removed.data

    def is_empty(self) -> None:
        return self.size == 0

    def __str__(self) -> str:
        if self.is_empty():
            return 'Deque: []'

        output = ''

        current = self.head.next
        for _ in range(self.size - 1):
            output += f'{current.data}, '
            current = current.next

        return f'Deque: [{output}{self.tail.prev.data}]'


if __name__ == '__main__':
    main()
